File: Problem 015 - Lattice paths.cpp
// Find the number of paths through a grid, being only allowed to go right or down.
#include <iostream>

#define GRIDHEIGHT 20
#define GRIDWIDTH 20

using namespace std;

// Array we will store intermediate results in to speed up execution
long long results[GRIDHEIGHT][GRIDWIDTH] = {0};

// Recursive function that returns our results
long long Lattice(int height, int width){
    // If height or width is zero there is only one path. Return 1.
    if (height == 0 || width == 0){
        return 1;
    }
    // Otherwise, if we already have a stored value for this width/height combination, return that.
    else if (results[height-1][width-1] != 0){
        return results[height-1][width-1];
    }
    // Otherwise we compute the value for this width/height combination, store it, then return it.
    else{
        long long r_val = Lattice(height -1, width) + Lattice(height, width - 1);
        results[height-1][width-1] = r_val;
        return r_val;
    }
}

int main()
{
    long long result = Lattice(GRIDHEIGHT, GRIDWIDTH);
    cout<<"The total number of routes through a "<<GRIDHEIGHT<<"x"<<GRIDWIDTH<<" grid is "<<result;
    return 0;
}

File: LUA/Problem015 - Lattice paths.lua
--[[
Starting in the top left corner of a 2×2 grid, and only being able to move to the right and down, there are exactly 6 routes to the bottom right corner.

How many such routes are there through a 20×20 grid?
]]

-- height and width of our grid
gridheight = 20
gridwidth = 20

-- We'll store previously calculated paths in this table
grid = {}

-- Initialize the table
for i = 1, gridheight do
	grid[i] = {}
	for j = 1, gridwidth do
		grid[i][j] = 0
	end
end

-- Recursive function that will give us the number of paths through a grid
function Lattice(height, width)
	-- If width or height has reached 0, there is only 1 way to the end. Return 1.
	if width == 0 or height == 0 then
		return 1
	-- If we have a stored value for this width/height combination, return the stored value
	elseif grid[height][width] ~= 0 then
		return grid[height][width]
	end
	-- Otherwise, recursively call the Lattice function twice, with reduced height and reduced width respectively.
	result = Lattice(height-1, width) + Lattice(height, width-1)
	-- Then store the result of those function calls so we can skip future calculations
	grid[height][width] = result
	return result
end

-- Get our answer from the Lattice function, and write the answer
answer = Lattice(gridheight, gridwidth)
io.write(string.format("The number of paths through a %d by %d grid is %.0f", gridheight, gridwidth, answer))